### ISSUE TITLE
onefile cannot execute when /tmp is mounted with `noexec`

### ISSUE THREAD

**Issue by deajan** [Nuitka/Nuitka#2142]
Not really a nuitka issue actually, but might be worth to talk about it in order to improve nuitka's portability with onefile binaries.

I compiled an ARMv7 binary with nuitka, in order to run it on a synology NAS.

On first run of the binary, I got 
```
ash-4.4# ./npbackup.arm.bin
Error, couldn't launch child (exec).: Permission denied
Error, couldn't launch child (exec).: Permission denied
```
Note: this is not a typo, I got the error twice.
After having fiddled around a bit, I found that `/tmp` dir was mounted with `noexec`.

I could indeed run the nuitka compiled binary with:
```
ash-4.4# TMPDIR=/run ./npbackup.arm.bin
```

In order to make sure a onefile binary can always run, maybe the onefile bootstrap could should detect if the extracted content can be executed.

Best regards.

**Comment by kayhayen** [comment]
What is the alternative to `TMPDIR` for extracting and running an executable, then?!

**Comment by kayhayen** [comment]
Is there a `/var/tmp` on said system?

**Comment by deajan** [comment]
Yes, there is `/var/tmp`, and it is mounted as root without `noexec` obviously.
In my scripts, i use to have the following logic:

```
if [ -w /tmp ]; then
	RUN_DIR=/tmp
elif [ -w /var/tmp ]; then
	RUN_DIR=/var/tmp
elif [ -w /run ]; then
	RUN_DIR=/run
else
	RUN_DIR=.
fi
```

You flagged "needs_example".
What can I provide you ?

**Comment by deajan** [comment]
Here's the standard mounts on the synology RS816 I'm compiling for:
```
ash-4.4# mount
/dev/md0 on / type ext4 (rw,relatime,journal_checksum,data=ordered)
sysfs on /sys type sysfs (rw,nosuid,nodev,noexec,relatime)
proc on /proc type proc (rw,nosuid,nodev,noexec,relatime)
devtmpfs on /dev type devtmpfs (rw,nosuid,size=513860k,nr_inodes=128465,mode=755)
securityfs on /sys/kernel/security type securityfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /dev/shm type tmpfs (rw,nosuid,nodev)
devpts on /dev/pts type devpts (rw,nosuid,noexec,relatime,gid=5,mode=620)
tmpfs on /run type tmpfs (rw,nodev,mode=755)
tmpfs on /sys/fs/cgroup type tmpfs (ro,nosuid,nodev,noexec,mode=755)
cgroup on /sys/fs/cgroup/systemd type cgroup (rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd)
cgroup on /sys/fs/cgroup/synomonitor type cgroup (rw,nosuid,nodev,noexec,relatime,name=synomonitor)
cgroup on /sys/fs/cgroup/blkio type cgroup (rw,nosuid,nodev,noexec,relatime,blkio)
cgroup on /sys/fs/cgroup/cpuset type cgroup (rw,nosuid,nodev,noexec,relatime,cpuset)
cgroup on /sys/fs/cgroup/memory type cgroup (rw,nosuid,nodev,noexec,relatime,memory)
cgroup on /sys/fs/cgroup/devices type cgroup (rw,nosuid,nodev,noexec,relatime,devices)
cgroup on /sys/fs/cgroup/cpu type cgroup (rw,nosuid,nodev,noexec,relatime,cpu)
cgroup on /sys/fs/cgroup/cpuacct type cgroup (rw,nosuid,nodev,noexec,relatime,cpuacct)
cgroup on /sys/fs/cgroup/freezer type cgroup (rw,nosuid,nodev,noexec,relatime,freezer)
debugfs on /sys/kernel/debug type debugfs (rw,nosuid,nodev,noexec,relatime)
tmpfs on /tmp type tmpfs (rw,nosuid,nodev,noexec)
configfs on /sys/kernel/config type configfs (rw,nosuid,nodev,noexec,relatime)
devtmpfs on /proc/bus/usb type devtmpfs (rw,nosuid,size=513860k,nr_inodes=128465,mode=755)
tmpfs on /run/user/196791 type tmpfs (rw,nosuid,nodev,relatime,size=103144k,mode=700,uid=196791,gid=196791)
/dev/mapper/vg1-volume_1 on /volume1 type ext4 (rw,nodev,relatime,journal_checksum,synoacl,stripe=48,jqfmt=vfsv0,usrjquota=aquota.user,grpjquota=aquota.group)
/dev/mapper/vg1-volume_2 on /volume2 type ext4 (rw,nodev,relatime,journal_checksum,synoacl,stripe=48,jqfmt=vfsv0,usrjquota=aquota.user,grpjquota=aquota.group)
none on /config type configfs (rw,nosuid,nodev,noexec,relatime)
```

**Comment by kayhayen** [comment]
Well, you can always use `--onefile-temp-spec` and not use the temp folder, but e.g. a user cache folder, and if you want things to delete, just use a volatile value like `%PID%` in there, otherwise use cached mode. I do not like using `/run` really, that must be a bad idea. It probably means you are root, and then `/anything` will do equally well.

**Comment by deajan** [comment]
For instance, I'm compiling the same program for windows-i386, windows-amd64, linux-amd64, linux-arm and linux-arm64.
For my ARM platform tests, I'm running the binary on a synology, but I cannot have guarantees that other linux-arm hosts don't have other `noexec` mountpoints.
Since I want to widely distribute my program, using `--onefile-temp-spec` would make a hard assumption that all hosts of a specific platform will share the same tmp folder layout.
Hence, it would be nice if Nuitka would have some fallback solution in case the TMPDIR folder is mounted with `noexec`.

**Comment by kayhayen** [comment]
I agree, that would be sweet. In fact, executing binaries of course does not require the x bit at all. We could just as well load it and execute it manually, but that is probably overkill.

**Comment by kayhayen** [comment]
Well, you can always use `--onefile-temp-spec` and not use the temp folder, but e.g. a user cache folder, and if you want things to delete, just use a volatile value like `%PID%` in there, otherwise use cached mode. I do not like using `/run` really, that must be a bad idea. It probably means you are root, and then `/anything` will do equally well.

**Comment by deajan** [comment]
> I agree, that would be sweet. In fact, executing binaries of course does not require the x bit at all. We could just as well load it and execute it manually, but that is probably overkill.

In my specific case, I have binaries that are shipped and executed by the nuitka compiled program, so executing the onefile content without x bit would work, but not in all cases.
TBH, I have no "good and easy" idea to know if a filesystem is mounted with `noexec`.
So unless a lot of code would be involved and not guarantee to work on all POSIX platforms, the safest solution would be to retry decompressing onefile payload on another path if not executable in the first place.

**Comment by kayhayen** [comment]
I think whatever `mount` does, we can do too of course. Querying file system and flags is surely simple API. Having a fallback to something is useful. I just hate `/run` and wonder if there is not any better way. Maybe like `$HOME/tmp` or something like that.

**Comment by kayhayen** [comment]
If we made it such, that the spec was a list of things, that are tried, that would allow to specifiy such ugly things as `/run` and have it work as a sort of fallback. That would require some sort of separator or list handling there, pretty annoying to code that all the way through as a list. But on the other hand, int and str, and supported define types, so why would list not be one of them as well.

Without something as boost, having loops out of defines, is going to be painful though. But I can see some general value in that, but honestly, I am overloaded right now.

**Comment by kayhayen** [comment]
I don't see myself doing this anytime soon, unless a commercial sponsor level subscriber asks for it.

**Comment by kayhayen** [comment]
Nothing seems to be forthcoming to fund this, I will wait some longer, but ultimately this might have to be closed.
